### PR TITLE
Allow using distance measure in the geo context precision

### LINKED ISSUE #24807: Error using miles for percision in Geo Location Context
**Elasticsearch version**:
5.1.2

**Description of the problem including expected versus actual behavior**:
Here is the search I want to run:

```
POST stores_search/_search?
{
  "suggest": {
    "store-suggest": {
      "prefix": "Combs",
      "completion": {
        "field": "suggest",
        "fuzzy": {
          "fuzziness": "AUTO",
          "transpositions": true
        },
        "size": 100,
        "contexts": {
          "location": [
            {
              "lat": 47.6062,
              "lon": -122.3321,
              "precision": "100m"
            }
          ]
        }
      }
    }
  }
}
```

I need to provide my precision in distance. The docs state this is possible:
```
The precision of the geohash to encode the query geo point. This can be specified as a distance value (5m, 10km etc.), or as a raw geohash precision (1..12). Defaults to index time precision level.
```

But instead of getting stores within that geo location, I get this error response:
```
{
  "error": {
    "root_cause": [
      {
        "type": "parsing_exception",
        "reason": "[geo] failed to parse field [precision]",
        "line": 1,
        "col": 57
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query_fetch",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "stores_search",
        "node": "Se4Iz_2cS4OGdz33kndgGg",
        "reason": {
          "type": "parsing_exception",
          "reason": "[geo] failed to parse field [precision]",
          "line": 1,
          "col": 57,
          "caused_by": {
            "type": "number_format_exception",
            "reason": "For input string: \"1000mi\""
          }
        }
      }
    ],
    "caused_by": {
      "type": "parsing_exception",
      "reason": "[geo] failed to parse field [precision]",
      "line": 1,
      "col": 57,
      "caused_by": {
        "type": "number_format_exception",
        "reason": "For input string: \"1000mi\""
      }
    }
  },
  "status": 400
}
```

If i change precision to `1` (or any number in (1..12)), my search completes with no errors.

**Steps to reproduce**:

 1. Create a mapping as an example try:
```
PUT place
{
    "mappings": {
        "shops" : {
            "properties" : {
                "suggest" : {
                    "type" : "completion",
                    "contexts": [
                        { 
                            "name": "place_type",
                            "type": "category",
                            "path": "cat"
                        },
                        { 
                            "name": "location",
                            "type": "geo",
                            "precision": 4
                        }
                    ]
                }
            }
        }
    }
}
```
 2. Put some documents in the new index, as an example:
```
PUT place/shops/1
{
    "suggest": {
        "input": "timmy's",
        "contexts": {
            "location": [
                {
                    "lat": 43.6624803,
                    "lon": -79.3863353
                },
                {
                    "lat": 43.6624718,
                    "lon": -79.3873227
                }
            ]
        }
    }
}
```
 3. Do a lil' query:
```
POST place/_suggest?pretty
{
    "suggest" : {
        "prefix" : "tim",
        "completion" : {
            "field" : "suggest",
            "size": 10,
            "contexts": {
                "location": [ 
                    {
                        "lat": 43.6624803,
                        "lon": -79.3863353,
                        "precision": "2m"
                    },
                    {
                        "context": {
                            "lat": 43.6624803,
                            "lon": -79.3863353
                        },
                        "boost": 2
                    }
                 ]
            }
        }
    }
}
```

result:
```
{
  "error": {
    "root_cause": [
      {
        "type": "parsing_exception",
        "reason": "[geo] failed to parse field [precision]",
        "line": 1,
        "col": 62
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "place",
        "node": "Se4Iz_2cS4OGdz33kndgGg",
        "reason": {
          "type": "parsing_exception",
          "reason": "[geo] failed to parse field [precision]",
          "line": 1,
          "col": 62,
          "caused_by": {
            "type": "number_format_exception",
            "reason": "For input string: \"2m\""
          }
        }
      }
    ],
    "caused_by": {
      "type": "parsing_exception",
      "reason": "[geo] failed to parse field [precision]",
      "line": 1,
      "col": 62,
      "caused_by": {
        "type": "number_format_exception",
        "reason": "For input string: \"2m\""
      }
    }
  },
  "status": 400
}
```

**Provide logs (if relevant)**:

### PR DESCRIPTION
**Pull request by imotov**
Adds support for distance measure, such as "4km", "5m" in the precision
field of the geo location context in context suggesters.

Fixes #24807

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/common/geo/GeoUtilTests.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    private int parsePrecision(CheckedConsumer<XContentBuilder, IOException> tokenGenerator) throws IOException {
+        XContentBuilder builder = jsonBuilder().startObject();
+        tokenGenerator.accept(builder);
+        builder.endObject();
+        XContentParser parser = createParser(JsonXContent.jsonXContent, BytesReference.bytes(builder));
+        parser.nextToken(); // {
+        parser.nextToken(); // field name
+        parser.nextToken(); // field value
+        return GeoUtils.parsePrecision(parser);
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/common/geo/GeoUtilTests.java]: Just for completeness, can we check that the current token pointer after parsing is as expected, e.g. by consuming the remaining tokens and asserting that they are closing elements or something like that?

File: server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 2]
```diff
@@ -459,6 +459,44 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
...
+                    throw e2;
+                } catch (IllegalArgumentException e3) {
+                    // this happens when distance too small, so precision > 12. We'd like to see the original string
+                    throw new IllegalArgumentException("precision too high [" + precision + "]", e3);
+                }
+            }
+        }
+    }
+
+    public static int checkPrecisionRange(int precision) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Can you add a short javadoc?

**Review by cbuescher — APPROVED** [review]
@imotov looks good to me, left a some minor comment that don't really need another review once adressed.

File: server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 3]
```diff
@@ -459,6 +459,50 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
         }
     }
 
+    /**
+     * Parse a precision that can be expressed as an integer or a distance measure like "1km", "10m".
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Maybe mention that this "precision" is the geohash length?

File: server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 4]
```diff
@@ -459,6 +459,50 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
...
+     */
+    public static int parsePrecision(XContentParser parser) throws IOException, ElasticsearchParseException {
+        XContentParser.Token token = parser.currentToken();
+        if (token.equals(XContentParser.Token.VALUE_NUMBER)) {
+            return XContentMapValues.nodeIntegerValue(parser.intValue());
+        } else {
+            String precision = parser.text();
+            try {
+                // we want to treat simple integer strings as precision levels, not distances
+                return XContentMapValues.nodeIntegerValue(Integer.parseInt(precision));
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: you should call either parseInt or nodeIntegerValue, not both?

File: server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 5]
```diff
@@ -459,6 +459,50 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
...
+                // try to parse as a distance value
+                try {
+                    return checkPrecisionRange(GeoUtils.geoHashLevelsForPrecision(precision));
+                } catch (NumberFormatException e2) {
+                    // can happen when distance unit is unknown, in this case we simply want to know the reason
+                    throw e2;
+                } catch (IllegalArgumentException e3) {
+                    // this happens when distance too small, so precision > 12. We'd like to see the original string
+                    throw new IllegalArgumentException("precision too high [" + precision + "]", e3);
+                }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: should we catch separately the call to checkPrecisionRange and geoHashLevelsForPrecision?

**Review by jpountz — APPROVED** [review]
I know you mostly moved some code around but I left some suggestions for improvements. Looks good to me regardless.

**Review by jpountz — APPROVED** [review]

**Pull request merged**